### ISSUE TITLE
 Support AES256/CBC/PKCS7Padding ???

### ISSUE THREAD

**Issue by VincentChen1212** [BruceWind/AESJniEncrypt#33]
HI

This is a good Project.

and
Is this Project Support AES256/CBC/PKCS7Padding ???
or something can reference ??

Thanks

**Comment by BruceWind** [comment]
Thanx for you issue.
Im going to add some algorithms in few weeks. 
I have a lot of plan ,but Im too lazy. sorry.

**Comment by BruceWind** [comment]
Hi ,man:
im wasnt lazy today. i made the repo -> [https://github.com/BruceWind/CryptoPPInNDK](https://github.com/BruceWind/CryptoPPInNDK).

some algoritms will soon.

**Comment by VincentChen1212** [comment]
Hi my Hero,

Many Many Many Many Many Thanks, I Very Need AES256/CBC/PKCS7Padding with C, In all the Github, I can't find any repo

Thanks you again

**Comment by BruceWind** [comment]
you are welcome!
Im busy this week, maybe you can to  use the [repo](https://github.com/BruceWind/CryptoPPInNDK), and add some algorithms. 
write some code test, build a so file, and  push to device.

Then you can  add you code to [this repo](https://github.com/BruceWind/AESJniEncrypt).

**Comment by BruceWind** [comment]
There is a great article, [C++使用AES+Base64算法对文本进行加密](https://blog.csdn.net/csdn49532/article/details/50686222),but the algoritm is AES128,isnt 256。

**Comment by VincentChen1212** [comment]
Hi,

Thanks for your helping.
My Business Model is using AES256/CBC/PKCS7Padding,
And I must use kotlin-native platform to build my project,
it only support c (don't support c++),
so I am so sad to find suitable project (I don't understand c/c++ )

Thanks You VERY VERY VERY Much

**Comment by BruceWind** [comment]
I think,  no kotlin-native in Android. Only "Java call native" in Android. Java can call native interface,regardless C , C++ or Assembly.  
Do you know I mean?

**Comment by BruceWind** [comment]
I mean you can use **extern "C" ** to call you C++ code. like this [calling-c-dll-from-java](https://stackoverflow.com/questions/9485896/calling-c-dll-from-java).

**Comment by VincentChen1212** [comment]
Hi,

Thanks for your helping.
My present situation : already create base64/sha/hmac using cinterop in kotlin-native project, now aes256 dosen't implement, but no suitable AES256/CBC/PKCS7Padding using c to reference.

Thanks You very much

**Comment by BruceWind** [comment]
okay,
this week, i dont have a lot of work, maybe the repo will go ahead a little.

**Comment by VincentChen1212** [comment]
Hi,
It's a Good news for me.
Thanks for your helping.

```
My Test Data is:
val data = "Discover interesting projects and people to populate your personal news feed"
val key = "XENT95ZAYCPH2CLFT7M2Y3WEWKB846B8"
val iv = "B3LCLM7CMXLY3ZAF"

The resule is:
JapXrfLPtd5bAkc69VtkeNlUeE/pgYdszgSQoqrDZEmg3GkA6wYx9+JaAtzNRg2ConKHmHtG/gCtNwpyO62OO+Jj+HnsxQwtvm8qkP2hpRI=
```

You are really my HERO

Thanks you very much

**Comment by BruceWind** [comment]
I think, maybe you are finding a wrong direction.


Java only provides PKCS#5 padding, but it is the same as PKCS#7 padding. See this question on Crypto.SE:

[What is the difference between PKCS#5 padding and PKCS#7 padding](https://crypto.stackexchange.com/questions/9043/what-is-the-difference-between-pkcs5-padding-and-pkcs7-padding)

They are interchangeable for the common block ciphers like AES and DES.
And such as [AES&DES algorithm in cnblogs](https://www.cnblogs.com/52fhy/p/6863133.html).



Then, I wrote a class to test.


```
import android.util.Base64;

import java.security.spec.AlgorithmParameterSpec;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

public class AESTest
{

    private static String KEY_AES = "D(G+KbPeShVmYq3t";
    public static String encrypt7(String value) {
        try {
            byte[] key = KEY_AES.getBytes("UTF-8");
            byte[] ivs = KEY_AES.getBytes("UTF-8");
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS7Padding");
            SecretKeySpec secretKeySpec = new SecretKeySpec(key, "AES");
            AlgorithmParameterSpec paramSpec = new IvParameterSpec(ivs);
            cipher.init(Cipher.ENCRYPT_MODE, secretKeySpec, paramSpec);
            return Base64.encodeToString(cipher.doFinal(value.getBytes("UTF-8")), Base64.DEFAULT);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static String encrypt5(String value) {
        try {
            byte[] key = KEY_AES.getBytes("UTF-8");
            byte[] ivs = KEY_AES.getBytes("UTF-8");
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            SecretKeySpec secretKeySpec = new SecretKeySpec(key, "AES");
            AlgorithmParameterSpec paramSpec = new IvParameterSpec(ivs);
            cipher.init(Cipher.ENCRYPT_MODE, secretKeySpec, paramSpec);
            return Base64.encodeToString(cipher.doFinal(value.getBytes("UTF-8")), Base64.DEFAULT);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
}
```
The  test result is equal.


Look this: [about padding in  crypto++ doc](https://www.cryptopp.com/docs/ref/struct_block_padding_scheme_def.html#abea06c498771e8f0ad0fbbc19416a979a622df395c2f0edc35f722d938faaad1f). I think, Crypto++ don't support PKCS7Padding.

**Comment by VincentChen1212** [comment]
Hi, 
Thanks for your helping.
I don't understand about C language, so if c only support pkcs5, it is ok.

My test data is use java ＆ C#, my coworker seem also use pkcs5 with java.

You are truly my savior.
I Really need C Code.

Thanks for your helping.

**Comment by VincentChen1212** [comment]
Hi, 

I Try my best to search on line
I found some pages
```https://www.cnblogs.com/cqvoip/p/8078790.html```
```http://www.longdw.com/aes-256-cbc-c-android-pkcs5padding/?utm_source=tuicool&utm_medium=referral```

So I try to Modify a little

I Add 4 function
```
a) aes256_encrypt_cbc_mdy (modify len for original function aes256_encrypt_cbc)
b) aes256_decrypt_cbc_mdy(modify len for original function aes256_decrypt_cbc)
c) aes256_encrypt_cbc_pkcs
d) aes256_decrypt_cbc_pkcs
```
c) and d) is for java to use

but now
if
in the c code with 『var padding=0』---> Only I Guess
the result will error
```
val data = "Discover interesting projects and people to populate your personal news feed"
val key = "XENT95ZAYCPH2CLFT7M2Y3WEWKB846B8"
val iv = "B3LCLM7CMXLY3ZAF"

Java：JapXrfLPtd5bAkc69VtkeNlUeE/pgYdszgSQoqrDZEmg3GkA6wYx9+JaAtzNRg2ConKHmHtG/gCtNwpyO62OO+Jj+HnsxQwtvm8qkP2hpRI=

Error Enc：JapXrfLPtd5bAkc69VtkeNlUeE/pgYdszgSQoqrDZEmg3Gk=
Error Dec：Discover interesting projects an
=======================================================

If My data append "1" in the front
val data = "1Discover interesting projects and people to populate your personal news feed"

Java：wgH+atRi3VAbSLqC91TJ5fcKju7fzh2cOeZxopxeKPBb22zol6rlhjZewZKlJnzsgBdVzkSsu4xXZeEuhhUjK2e7eJGYE6AN4rFhJXlxXpo=

Correct Enc：wgH+atRi3VAbSLqC91TJ5fcKju7fzh2cOeZxopxeKPBb22zol6rlhjZewZKlJnzsgBdVzkSsu4xXZeEuhhUjK2e7eJGYE6AN4rFhJXlxXpo=
Correct Dec：1Discover interesting projects and people to populate your personal news feed
```

Could you please me to find where is error.

Here is my full Code
```
https://drive.google.com/open?id=1-GJXStGxZ3pi1fVKPcZlK7wRauNbR5V2
```

Thanks for your helping.

**Comment by BruceWind** [comment]
Obviously,length of `Discover interesting projects an` is 32, 32 same to  16 * 2.
Perhaps, "length" and some loop have mistakes.

If you ask about your aes algorithm, this isn't a question and answer platform.Pls go to [Stackoverflow](https://stackoverflow.com). Im Sorry.

**Comment by BruceWind** [comment]
sorry, AES no longer support, 
I have pushed code that use chacah20 instead of AES. https://github.com/BruceWind/AESJniEncrypt/pull/40

**Comment by PeterleSoftwares** [comment]
> 
> 
> There is a great article, [C++使用AES+Base64算法对文本进行加密](https://blog.csdn.net/csdn49532/article/details/50686222),but the algoritm is AES128,isnt 256。

Thanks to mr. BruceWind!!! He saved my life with this article referral!!!

**Comment by BruceWind** [comment]
Very appreciate that this repo helped you. Thanks.
